src/cargo/ops/cargo_rustc/fingerprint.rs

```diff
@@ -57,6 +57,15 @@ pub fn prepare_target<'a, 'cfg>(cx: &mut Context<'a, 'cfg>,
     let compare = compare_old_fingerprint(&loc, &*fingerprint);
     log_compare(unit, &compare);
 
+    if compare.is_err() {
+        let source_id = unit.pkg.package_id().source_id();
+        let sources = cx.packages.sources();
+        let source = try!(sources.get(source_id).chain_error(|| {
+            internal("missing package source")
+        }));
+        try!(source.verify(unit.pkg.package_id()));
+    }
```

Comment:
Can you add a comment explaining what his new block is doing? I don't follow. It looks like it's recovering from an error by doing extra validation. In what situation does comparing the fingerprint produce an error, and in what situation does doing the extra `verify` call help and how?